### ISSUE TITLE
understanding stream issues using AnalyticsListener

### ISSUE THREAD

**Issue by ilker-aktuna** [google/ExoPlayer#8097]
I am using the analytics listener to create an app which will simulate the user experience and supply stats about some certain video streams.
The app  is actually ready and I am able to get events from AnalyticsListener.
However, my most important concern is to understand when the app is not successfully getting streams.
So I am focusing on loadcomplete and loaderror events.
Most of the time, when the streaming server has slight problems, the stream is not cut but slows down.
So I want to be able to understand the issues by counting loadcomplete events, because loaderror occurs very rarely.

However, loadcomplete dependds on the bandwidth, duration, chunks, bitrate.
So I am not able to guarantee a static threshold for the loadcomplete event.

Is there a way to guarantee a minimum loadcomplete event for a predefined duration of video ?

Or else, can I define a calculation method based on bitrate and duration ?

**Comment by andrewlewis** [comment]
> However, my most important concern is to understand when the app is not successfully getting streams.

If there is a transient network/server issue but it doesn't impact users (because there is enough media in memory to play out and 'absorb' the temporary delay in loading) do you want to detect that? If the goal is really to measure whether the user experience is good, you may want to focus on user-visible rebuffer events instead, rather than trying to come up with a deadline for chunk loads completing.

> Is there a way to guarantee a minimum loadcomplete event for a predefined duration of video ?
> Or else, can I define a calculation method based on bitrate and duration ?

I'm not sure how to answer this because there are many variables (as you note) and I'm not sure what the exact case you are trying to detect is.

For a given chunk load there is a deadline after which the player will rebuffer, but I think it's easiest to detect crossing that deadline by detecting the player transitioning to the buffering state.

You could inspect the `LoadEventInfo`s on successive load complete events to detect when the effective bandwidth dropped (or just poll the `BandwidthMeter`), but a drop in bandwidth could have many different causes.

@tonihei is the expert on analytics and may be able to give a more helpful answer.

**Comment by ilker-aktuna** [comment]
actually I observed issues with our transcoder and the real user experience is "freeze" in video for about 1-2 seconds.
My app does not report loaderror for this period.
But I see a change in loadcomplete (it decreases in general) 
So I would like to put a threshold (whewre I collect the data from my app) for loadcomplete events.
But as I wrote, I the loadcomplete is variable, I am not able to put a certain threshold to compare with.

I had thought maybe it would be possible to fix the loadcomplete events to a minimum number for normal conditions. (using loadcontrol and/or bandwidthmeter ??)
But I don't know how...

User visible rebuffer events you said. How can I get weird rebuffer events ?

**Comment by tonihei** [comment]
I think there is some misunderstanding about what `onLoadError` and `onLoadComplete` are reporting. 

Your media is loaded in chunks and each individual chunk load will get `onLoadStarted` and `onLoadCompleted`/`onLoadError`/`onLoadCanceled` events. This corresponds to the actual network operations and thus the frequency of these events can't be changed. Note that all this doesn't directly influence the user-visible playback in any way because ExoPlayer keeps a queue of media (up to 50 seconds by default) and the user-visible rendering happens at the other end of the queue.

From what I understand, you are probably interested in `onPlaybackStateChanged`. If the state changes from `Player.STATE_READY` to `Player.STATE_BUFFERING`, your user will observe a freeze or rebuffer. This is roughly when the mentioned media queue becomes empty.

**Comment by ilker-aktuna** [comment]
ok. then you recommend not to monitor "loaderror" but buffering events.
How can I monitor abnormal buffering ?
is there an event for that ?
Because normally it will buffer for the first load and that is totally normal, am I correct ?

**Comment by tonihei** [comment]
Yes, that's correct. You can just disregard the initial buffering state. 

If you don't need live monitoring of the playback but just a summary of what happened, you could also add an `PlaybackStatsListener` that will give you a list of more granular states, e.g. PLAYBACK_STATE_JOINING_FOREGROUND for the initial joining time, or PLAYBACK_STATE_BUFFERING for "abnormal buffering" as you called it.

**Comment by ilker-aktuna** [comment]
can you provide an example code for usage of this ?

**Comment by tonihei** [comment]
For live monitoring:
```
SimpleExoPlayer player = new SimpleExoPlayer.Builder(/*...*/).build();
player.addEventListener(new EventListener() {
  boolean hasSeenFirstBufferingState = false;

  @Override 
  public void onPlaybackStateChanged(@Player.State int playbackState) {
    if (playbackState == Player.STATE_BUFFERING) {
      if (hasSeenFirstBufferingState) {
        Log.d("MONITORING", "Now rebuffering!");
      }
      hasSeenFirstBufferingState = true;
    }
  }
});

```

For analytics purposes without live monitoring and with more detailed states:
```
/* init */
SimpleExoPlayer player = new SimpleExoPlayer.Builder(/*...*/).build();
PlaybackStatsListener playbackStatsListener = new PlaybackStatsListener(/* keepHistory= */ true, null);
player.addAnalyticsListener(playbackStatsListener);

/* much later after playback finished */
playbackStatsListener.finishAllSessions();
PlaybackStats stats = playbackStatsListener.getPlaybackStats();
for (EventTimeAndPlaybackState timeAndState : stats.playbackStateHistory) {
  Log.d("STATEHISTORY", timeAndState.eventTime.realtimeMs + ": " + timeAndState.playbackState);
}
```

**Comment by ilker-aktuna** [comment]
actually I am not interested in state history. I just want to get an idea of how the stream performed.
is this ok ?

To define:

```
private PlaybackStatsListener myPlaybackStatsListener;
        myPlaybackStatsListener=new PlaybackStatsListener(true, null);
        playerManager.exoPlayer.addAnalyticsListener(myPlaybackStatsListener);
```

after playback:
```
                        PlaybackStats myStats= myPlaybackStatsListener.getPlaybackStats();
                        Log.d("ISSUES", "total rebuffer:"+myStats.totalRebufferCount);
```

is "totalRebufferCount" what I must look at ?
what else ?

and how can I reset counters before starting a new session ?

what is "fatalErrorCount" ?

**Comment by tonihei** [comment]
Yes, that's the short version if you don't want the history. You can also use `new PlaybackStatsListener(/* keepHistory= */ false, null)` in this case because you don't need a history.

I can't answer what you'd like to look at because it depends on what you'd like to know. `totalRebufferCount` and `fatalErrorCount` are good metrics to track in general. Fatal errors is when the playback fails completely without continuing. 

> and how can I reset counters before starting a new session ?

This will happen automatically whenever you start playing a new `MediaItem` with the same player. If this is what your app is doing, you should use the `callback` variable that can be passed to the `PlaybackStatsListener` constructor. It will be called whenever a session is finished and all stats are final. Alternatively, you can get the overall stats (for all sessions together) by calling `playbackStatsListener.getCombinedPlaybackStats()`.

**Comment by ilker-aktuna** [comment]
I didn't understand the "callback" variable.
What should I use for this variable ? and how ?

**Comment by tonihei** [comment]
To get informed when the `PlaybackStats` for one playback session are ready. It's just a regular Java interface you can implement. Please also have a look at the JavaDoc to see how classes look like.

**Comment by ilker-aktuna** [comment]
well I actually get stats just before changing the media source. So I don't need that callback function. (I guess ?)
but it doesn't seem to reset the rebufferCount when mediasource is changed.
it was 0 initially, after some time it got 1 and it keeps 1 after that.

what do you mean by "when the PlaybackStats for one playback session are ready" ?
are they not live ?
I mean "updated when events occur".

well, if I need to use the callback, then I need to understand when it will really be called. 
because I post the stats (with all other collected stats) to a server (http call) 
Normally I post all stats right before I change the mediasource.
if the stats are reset at that time, I have to change where I post the stats.

**Comment by ilker-aktuna** [comment]
btw, there is "totalPauseBufferCount" and "totalReBufferCount" 
which one is useful for me ?

totalPauseBufferCount
The total number of times a playback has been paused while rebuffering.

totalRebufferCount
The total number of times a rebuffer occurred. This excludes initial joining and buffering after seek.

**Comment by tonihei** [comment]
> well I actually get stats just before changing the media source. So I don't need that callback function. (I guess ?)

That's correct, yes.

> but it doesn't seem to reset the rebufferCount when mediasource is changed.
it was 0 initially, after some time it got 1 and it keeps 1 after that.

Is that with `getCombinedPlaybackStats()` (expected) or `getPlaybackStats()` (unexpected). And how do you "change" the media source? 

> what do you mean by "when the PlaybackStats for one playback session are ready" ?are they not live ? I mean "updated when events occur".

They are updated every time you call `getPlaybackStats()` and you can call it as often as you like. The callback is helpful to know when the stats for one playback are final and don't change anymore.

> btw, there is "totalPauseBufferCount" and "totalReBufferCount" - which one is useful for me ?

That again depends on what you'd like to know. But I guess `totalRebufferCount` is more useful. `totalPauseBufferCount` counts how often the user paused playback while rebuffering, so it's more measuring user behaviour.

**Comment by ilker-aktuna** [comment]
I used getPlaybackStats() and I change the media with: 

```
        HlsMediaSource hlsMediaSource = new HlsMediaSource.Factory(dataSourceFactory).createMediaSource(hlsVideoUri);
        player.setPlayWhenReady(true);
        player.prepare(hlsMediaSource);

```

btw, to just test the callback and see when it is called, I changed my PlaybackStatsListener definition as below. However, I don'T see it called in the log.
Why ?

```
                myPlaybackStatsListener=new PlaybackStatsListener(true, new PlaybackStatsListener.Callback() {
                    @Override
                    public void onPlaybackStatsReady(AnalyticsListener.EventTime eventTime, PlaybackStats playbackStats) {
                        fatalErrorCount=playbackStats.fatalErrorCount;
                        rebufferCount=playbackStats.totalRebufferCount;
                        Log.d("POST:","stats ready");
                        Log.d("POST: ", ""+rebufferCount+":"+fatalErrorCount);
                    }
                });
                player.addAnalyticsListener(myPlaybackStatsListener);
```

**Comment by ilker-aktuna** [comment]
ok. now I see it being called.
and I see that the stats is reset when the callback is called.

actually , when the callback is called , the rebuffercount gets zero.
but when I reach the end of my test (about 15 seconds of playback) , rebuffercount  is 1.
this happens in all my tests. 
so is this the initial buffering ?
and, really when the callback is called ?

**Comment by ilker-aktuna** [comment]
last night we made several tests  with the transcoding platform.
We even closed the stream on cdn to create an issue which we could observe with the app.
However, there is a also a buffering mechanism on cdn and it does not drop the stream in a short time.

besides, I see 1 or 2 "rebuffer" on the app even when the stream is running fine.

So, this rebuffer state does not help me to identify problems.

What I catch with my app is the change in "loadcomplete"
so I want to go back to counting loadcomplete events.

But with that, I need to fix the number of loadcomplete to a number (for normal conditions) and then compare with that value to detect issues.
Is it possible to fix the loadcomplete events using LoadControl or BandwidthMeter ?

**Comment by ilker-aktuna** [comment]
for example,  can I define loadcontrol as below:
                DefaultLoadControl loadControl = new DefaultLoadControl.Builder().setPrioritizeTimeOverSizeThresholds(false).setBufferDurationsMs(1000, 1000, 1000, 1000).build();

and then , excpect a minimum of "loadcomplete" events based on stream bitrate ?

**Comment by andrewlewis** [comment]
It sounds like you are trying to identify a problem with your transcoder from the client, but isn't it better first to add monitoring in the transcoder itself, so you can eliminate the possibility of irrelevant communication channel and client issues?

> and then , excpect a minimum of "loadcomplete" events based on stream bitrate ?

As discussed earlier in this thread, load complete events are telling you when (for example) a media chunk was loaded. I think you can do some calculations yourself to figure out how many you expect based on your particular setup (for example, how long your media chunks are) if that really want you want to do.

As a general point it would be helpful if you could try to focus your questions to specific points about ExoPlayer APIs and do a bit more investigation yourself before replying. We have to balance time between responding to support issues like this and making improvements to ExoPlayer itself, and so we don't have time to go into lots of detail on issues that are specific to particular use cases.

**Comment by ilker-aktuna** [comment]
actually the whole purpose of this app that I am working on is to monitor the transcoding platform from user perspective.
it is not a real client. I am just developing this app to solely for this purpose of monitoring the stream from mobile device.

There's also a commercial quality monitoring platform that we use but it is on the core network backbone and is not getting results from user perspective. 
With this app I am also simulating the login process and then testing the stream.

> As discussed earlier in this thread, load complete events are telling you when (for example) a media chunk was loaded. I think you can do some calculations yourself to figure out how many you expect based on your particular setup (for example, how long your media chunks are) if that really want you want to do.

When I check the loadcomplete events I see that for a period of 30 seconds playback, under normal conditions (transcoding platform performing well) , 6 to 8 loadcomplete events occur. But half of them are like empty (difference of start and end time is zero) , the other half is 5-6 seconds each.
What does that mean ?
Why do I observe zero length chunks when everything is normal ?

Also, why do I see 1-2 rebuffer events when stream is good ?

> As a general point it would be helpful if you could try to focus your questions to specific points about ExoPlayer APIs and do a bit more investigation yourself before replying. We have to balance time between responding to support issues like this and making improvements to ExoPlayer itself, and so we don't have time to go into lots of detail on issues that are specific to particular use cases.

Actually this is what I am trying to do. I am trying to understand how I can configure Exoplayer (using loadcontrol , bandwidthmeter maybe) to get a stable "normal condition" count of events to define my thresholds.
Although it is nice to have features like AnalyticsListener , there seems to be very few documentation and examples about usage of advanced Exoplayer features.

I am sorry for taking your time for some purpose that you don't believe in...

**Comment by ilker-aktuna** [comment]
Hi,

Sorry for going back to this issue. But there is a weird situation.

Today there was a problem on our streaming platform.
My app currently looks for loadcomplete and loaderror events. It reports loadcomplete events with duration and with 0 duration separately.
It also reports count of rebuffer events.

At the time of problem, loadcomplete events increased (both zero duration and normal loadcomplete events increased)
rebuffer events were NOT increased. loaderror events were NOT increased.

So, what does this mean ? only loadcomplete events were increased at the time of problem.

If this is caused by rebuffer, then why the rebuffer events were not increased ?

Is there a way of disabling rebuffer totally ?
This way maybe I can just count on "loadcomplete" events. When they decrease, I can say that there is a problem.

**Comment by ilker-aktuna** [comment]
is it possible to limit the chunks downloaded by setting a fixed-duration for media to be played ?
my media is Hls accessed with a m8u playlist url.

```
            hlsVideoUri = Uri.parse(currURL);
        DataSource.Factory dataSourceFactory = new DefaultDataSourceFactory(this,
                Util.getUserAgent(this, "Test"));
        HlsMediaSource hlsMediaSource =
                new HlsMediaSource.Factory(dataSourceFactory)   //.setAllowChunklessPreparation(true)
                        .createMediaSource(MediaItem.fromUri(hlsVideoUri));
        player.setPlayWhenReady(true);
        player.prepare(hlsMediaSource);
```

**Comment by andrewlewis** [comment]
> When I check the loadcomplete events I see that for a period of 30 seconds playback, under normal conditions (transcoding platform performing well) , 6 to 8 loadcomplete events occur. But half of them are like empty (difference of start and end time is zero) , the other half is 5-6 seconds each.
> What does that mean ?
> Why do I observe zero length chunks when everything is normal ?

I'm guessing the empty ones may be requests that are not media chunks (initialization data for example). Can you tell from the load event info or media load data passed to `AnalyticsListener.onLoadCompleted` what kind of loads they were?

> Also, why do I see 1-2 rebuffer events when stream is good ?

How are you determining that the stream is good? If the player is rebuffering then probably the media wasn't loading fast enough (that doesn't necessarily imply that the chunks weren't available).

> I am sorry for taking your time for some purpose that you don't believe in...

I don't mean to imply I don't believe in it but it's just a question of needing to prioritize spending time on issues that provide significant benefit to all users.
> Is there a way of disabling rebuffer totally ?

You could try increasing the player's buffer buffer for playback to be equal to the duration of the media, via `DefaultLoadControl`. Then the player won't transition out of the buffering state until it's loaded the entire stream. I think that any setup where you have a smaller buffer leaves open the possibility that loading will become slow and playback will exhaust all loaded data. Of course this isn't appropriate for a real playback application as it may use lots of RAM and the user will need to wait until the entire stream is in memory before playback starts.

For the rest of your questions I don't think there's enough information to give proper answers. With information you have about how you are producing the stream being played by the client, you should be able to figure it out. The only other thing I can think of to help is to explain that when the player rebuffers (changing from ready to buffering) it generally means it has exhausted its buffer of loaded compressed data (for example, we rendered all the video frames we have buffered and loading is ongoing but no frames are arriving because it's slow); and on load completed just means that we made a request for a chunk of media from the server and loading it has finished.

I am going to close the issue now but if you have any specific problems where you're able to provide sufficient information for us to help please feel free to file a new issue.

**Comment by ilker-aktuna** [comment]
you have questions to me but you closed the thread. how should I proceed for the answers now ?

**Comment by ilker-aktuna** [comment]
>>> Why do I observe zero length chunks when everything is normal ?

>> I'm guessing the empty ones may be requests that are not media chunks (initialization data for example). Can you tell from the load event info or media load data passed to AnalyticsListener.onLoadCompleted what kind of loads they were?

how can I determine the type of loadcomplete events ?

>>> Is there a way of disabling rebuffer totally ?

>> You could try increasing the player's buffer buffer for playback to be equal to the duration of the media, via DefaultLoadControl. Then the player won't transition out of the buffering state until it's loaded the entire stream. I think that any setup where you have a smaller buffer leaves open the possibility that loading will become slow and playback will exhaust all loaded data. Of course this isn't appropriate for a real playback application as it may use lots of RAM and the user will need to wait until the entire stream is in memory before playback starts.

how can I set the player's buffer to a static duration and stop the media after that duration ?
I really want to play just a 30 seconds of live media and then stop.
I only want to analyze if 30 seconds of media is well received from the server. 
so : start media - get 30 seconds of media - stop
how can I achieve that ?

at time of problem, I observe that loadcomplete events increase but rebuffer does not increase. What does that mean ?

**Comment by andrewlewis** [comment]
> how can I determine the type of loadcomplete events ?

Please look at the load event info and media load data information, as mentioned in my previous comment.

> so : start media - get 30 seconds of media - stop
> how can I achieve that ?

See https://exoplayer.dev/media-items.html#clipping-a-media-stream.

> at time of problem, I observe that loadcomplete events increase but rebuffer does not increase. What does that mean ?

I don't have enough information to help I'm afraid, as I said in the earlier comment.

**Comment by ilker-aktuna** [comment]
ok.
for a live stream from m8u URL, what should be the startPositionMs and endPositionMs ?

**Comment by ilker-aktuna** [comment]
when I log the mediaLoadData.dataType , I see that it reports dataType 1 for events with duration, dataType 4 for events with no duration.
What is datatype 4 ?
and what can I get from loadEventInfo to analyze it further ?

how can I set the player's buffer to a static duration and stop the media after that duration ?
you told me to check:
https://exoplayer.dev/media-items.html#clipping-a-media-stream

so I set it like:
        MediaItem mediaItem = new MediaItem.Builder()
                .setUri(hlsVideoUri)
                .setClipStartPositionMs(0)
                .setClipEndPositionMs(30000)
                .build();
        HlsMediaSource hlsMediaSource =
                new HlsMediaSource.Factory(dataSourceFactory)   //.setAllowChunklessPreparation(true)
                        .createMediaSource(mediaItem);

but it still gets about 6-8 chunks per 30 seconds.

**Comment by ojw28** [comment]
> What is datatype 4 ?

This is easy to find out for yourself by reading `dataType` field's Javadoc:

> /** One of the {@link C} {@code DATA_TYPE_*} constants defining the type of data. */

And then looking at the various `C.DATA_TYPE_` constants. `C.DATA_TYPE_MANIFEST` is the one that has a value of 4. For HLS this means it's a master or media playlist file.

> for a live stream from m8u URL, what should be the startPositionMs and endPositionMs ?

If you mean a media playlist, `mediaStartTimeMs` and `mediaEndTimeMs` will both be set to `C.TIME_UNSET`. This is also quite easy for you to work out yourself by looking at what's being reported via `AnalyticsListener`. For example, you could implement `onLoadCompleted` to log a bunch of values from the reported objects, and then just look at what these fields are set to when the URL is for a media playlist.

> and what can I get from loadEventInfo to analyze it further ?

If you're asking what else is in `LoadEventInfo`, you could attach a debugger and have a look. Or browse the source code, or read the Javadoc.

I appreciate the task you're working on is probably fairly difficult, but it's really not the purpose of this issue tracker to provide per-developer consultation, particularly when many of the questions being asked are things that could be answered without assistance with a bit of research. Our team doesn't scale to do this. Closing this issue again. If you have any _specific_ issues with ExoPlayer (i.e., you think we're reporting something incorrectly via `AnalyticsListener`, have properly researched it, and have reproducible steps we can follow), then please let us know (you can do so here, even though the issue is marked as closed, we'll still see it). Thanks.

**Comment by ilker-aktuna** [comment]
actually you are getting it totally wrong.

without checking back the previous discussion you are judging and trying to answer.

"for a live stream from m8u URL, what should be the startPositionMs and endPositionMs ?"

I asked this because I asked how I could get a predefined duration of clip from the stream and I was recommended to check:
https://exoplayer.dev/media-items.html#clipping-a-media-stream

But I need to give 2 parameters : startPositionMs  and endPositionMs 
So , if I am trying to get a live stream which is being filled up live, how can I set these parameters to get the next (or maybe last) 30 seconds.
I still couldn't figure it out after your explanation because you are not focusing on what I have asked. You are trying to find my mistake. 
Well, I was told to use this issue tracker and there is an option for "questions" rather than bugs of feature requests. So I used it.
I had thought that "questions" would be answered with a lower priority.  
If you don't approve such questions, then I won't use it again for my questions. 

"And then looking at the various C.DATA_TYPE_ constants. C.DATA_TYPE_MANIFEST is the one that has a value of 4. For HLS this means it's a master or media playlist file."
Well, I am not an HLS or  even streaming expert. So this information is useless for me.
You have a prejudice that I did not check Javadoc , but I did. However the information on Javadoc is not really helping me.


>>>and what can I get from loadEventInfo to analyze it further ?

>>If you're asking what else is in LoadEventInfo, you could attach a debugger and have a look. Or browse the source code, or read the Javadoc.

I tried to explain what I am trying to achieve.
I was told that loadcomplete is not meaningful, instead  using rebuffer counts was recommended.
However, I see no change in rebuffer counts at the time of a real issue.
So I am still trying to find a way.

At first, I had thought that AnalyticsListener was introduced to analyze issues with the stream. And that was typically what I was looking for.
Aparently, it was not meant to be that way , or this feature is not being recommended for this purpose. 

Thanks for all people who have tried to helped. Unfortunately none of the recommendations addressed what I have asked for.
